system_test/aest_hard_fork_SUITE.erl

```diff
@@ -0,0 +1,564 @@
+    wait_for_height_syncing(HeightToBeMinedWithNewProtocol, [new_node4], {{45000, ms}, {1000, blocks}}, Cfg),
+    ?assertEqual(HashMined, maps:get(hash, get_block_by_height(new_node4, HeightToBeMinedWithNewProtocol, Cfg))),
+    aest_nodes:kill_node(new_node4, Cfg),
+    aest_nodes:kill_node(new_node3, Cfg),
+    ok.
+
+%=== INTERNAL FUNCTIONS ========================================================
+
+db_backup_tar(Cfg) ->
+    {_, DataDir} = proplists:lookup(data_dir, Cfg),
+    filename:join(DataDir, ct:get_config(db_backup_tar)).
+
+start_node(NodeName, Cfg) ->
+    aest_nodes:start_node(NodeName, Cfg),
+    F = fun() -> try get_block_by_height(NodeName, 0, Cfg), true catch _:_ -> false end end,
+    aec_test_utils:wait_for_it(F, true),
+    ct:log("Node ~p started as version~n~p", [NodeName, get_version(NodeName, Cfg)]),
+    ok.
+
+restore_db_backup_on_node(NodeName, Cfg) ->
+    {ok, TarBin} = file:read_file(db_backup_tar(Cfg)),
+    restore_db_backup_on_node(
+      NodeName,
+      TarBin, ct:get_config(db_backup_content),
+      ?DB_BACKUP_DEST_DIR,
+      Cfg).
+
+restore_db_backup_on_node(NodeName, TarBin, Content, DestDir, Cfg) ->
+    ct:log("Restoring DB backup of byte size ~p on node ~s",
+           [byte_size(TarBin), NodeName]),
+    _ = aest_nodes:run_cmd_in_node_dir(NodeName, ["mkdir", DestDir], Cfg),
+    "" = aest_nodes:run_cmd_in_node_dir(NodeName, ["ls", DestDir], Cfg),
+    ok = aest_nodes:extract_archive(NodeName, DestDir, TarBin, Cfg),
+    Content = aest_nodes:run_cmd_in_node_dir(NodeName, ["ls", DestDir], Cfg),
+    Dest = DestDir ++ "/" ++ Content,
+    ErlCmd =
+        "{atomic, [_|_] = Tabs} = mnesia:restore(\"" ++ Dest ++ "\", []), ok.",
+    run_erl_cmd_on_node(NodeName, ErlCmd, "ok", Cfg),
+    Top = #{height := TopHeight,
+            hash := TopHash} = aest_nodes:get_top(NodeName, Cfg),
+    ct:log("Restored DB backup on node ~s, whose top is now~n~p",
+           [NodeName, Top]),
+    {ok, {TopHash, TopHeight}}.
+
+mock_pow_on_node(NodeName, Cfg) ->
+    S =
+        "-module(aec_pow_cuckoo). "
+        "-export([generate/3, verify/4]). "
+        "generate(_, _, Nonce) -> Evd = lists:duplicate(42, 0), {ok, {Nonce, Evd}}. "
+        "verify(_,_,_,_) -> true.",
+    load_module_on_node(NodeName, aec_pow_cuckoo, S, Cfg).
+
+load_module_on_node(NodeName, Module, String, Cfg) ->
+    ct:log("Module ~s:~n~s", [Module, String]),
+    Tokens = dot_ending_token_lists(String),
+    ct:log("Tokens:~n~p", [Tokens]),
+    Forms = to_forms(Tokens),
+    ct:log("Forms:~n~p", [Forms]),
+    {ok, Module, Binary, []} = compile:forms(Forms, [return_errors,
+                                                     return_warnings]),
+    ErlCmd =
+        lists:flatten(
+          io_lib:format(
+            "{module, _} = code:load_binary(~s, \"Dummy Filename\", ~w), ok.",
+            [Module, Binary])),
+    run_erl_cmd_on_node(NodeName, ErlCmd, "ok", Cfg),
+    ok.
```

Comment:
Instead of defining and compiling code in code, I would rather see a mocks folder with some Erlang modules in it under the system test data directory. This isn't as easy to maintain.